app/controllers/flood_risk_engine/enrollments/addresses_controller.rb

```diff
@@ -0,0 +1,107 @@
+require_dependency "flood_risk_engine/application_controller"
+
+module FloodRiskEngine
+  module Enrollments
+    class AddressesController < ApplicationController
+
+      def new
+        build_address
+        check_for_errors
+      end
+
+      def create
+        build_address
+        update
+      end
+
+      def edit
+        check_for_errors
+      end
+
+      def update
+        clear_error_params
+        redirect_to target_url
+      end
+
+      private
+
+      def form
+        @form ||= AddressForm.new(enrollment, address)
+      end
+      helper_method :form
+
+      def target_url
+        if save_form!
+          step_forward
+          enrollment_step_path(enrollment, enrollment.current_step)
+        else
+          session[:error_params] = {
+            address: params[:address]
+          }
+          failure_url
+        end
+      end
+
+      def failure_url
+        if params[:action] == "create"
+          new_enrollment_address_path(
+            enrollment,
+            address_initial_attributes.merge(
+              check_for_error: true,
+              step_back_to_postcode: params[:step_back_to_postcode]
+            )
+          )
+        else
+          edit_enrollment_address_path(
+            enrollment,
+            address,
+            check_for_error: true
+          )
+        end
+      end
+
+      def save_form!
```

Comment:
Consider this example:
```ruby
class Plus
  def modify(x, y)
    x + y
  end
end

class Minus
  include MinusFoo
end

module MinusFoo
  def modify(x, y)
    x - y
  end
end 
```
If I want to understand how `Plus` works, I only need to look at the `Plus` code. 

If I want to work out how `Minus` works, I need to look at both the `Minus` code and the `MinusFoo` code.

Abstraction is often good, and can make code more readable and easier to follow. But if done too much or too soon, it can make code harder to follow.

If the MinusFoo code needs to be re-used more than twice, the case for abstraction changes. For example if we added:
```
class TakeAway
  include MinusFoo
end

class Subtract
  include MinusFoo
end
```
The case for MinusFoo is now compelling.

At the moment, I think abstraction will make the addresses controller code harder to follow, without adding a significant benefit.

If more controllers are added, then I think we should look at abstraction.